src/rust/engine/grpc_util/src/tls.rs

```diff
@@ -33,71 +49,105 @@ impl TryFrom<Config> for ClientConfig {
+                format!(
                 "Could not discover root CA cert files to use TLS with remote caching and remote \
             execution. Consider setting `--remote-ca-certs-path` instead to explicitly point to \
             the correct PEM file.\n\n{err}",
               )
-            })?;
-            let mut root_cert_store = RootCertStore::empty();
-            for cert in native_root_certs {
-              root_cert_store.add_parsable_certificates(&[cert.0]);
+              })?;
+
+              for cert in native_root_certs {
+                root_cert_store.add_parsable_certificates(&[cert.0]);
+              }
             }
-            tls_config.with_root_certificates(root_cert_store)
           }
+
+          tls_config.with_root_certificates(root_cert_store)
         };
 
-        if let Some(MtlsConfig { key, cert_chain }) = config.mtls {
+        if let Some(MtlsConfig { cert_chain, key }) = config.mtls {
           tls_config
-            .with_client_auth_cert(
-              cert_chain_from_pem_bytes(cert_chain)?,
-              der_key_from_pem_bytes(key)?,
-            )
-            .map_err(|err| format!("Error creating MTLS config: {:?}", err))?
+            .with_client_auth_cert(cert_chain, key)
+            .map_err(|err| format!("Error creating mTLS config: {:?}", err))?
         } else {
           tls_config.with_no_client_auth()
         }
       }
     };
 
-    tls_config.alpn_protocols = vec![b"h2".to_vec()];
     Ok(tls_config)
   }
 }
 
 #[derive(Clone)]
 pub struct MtlsConfig {
-  /// PEM bytes of the private key used for MTLS.
-  pub key: Vec<u8>,
-  /// PEM bytes of the certificate used for MTLS.
-  pub cert_chain: Vec<u8>,
+  /// DER bytes of the private key used for mTLS.
+  pub key: rustls::PrivateKey,
+  /// DER bytes of the certificate used for mTLS.
+  pub cert_chain: Vec<rustls::Certificate>,
+}
+
+impl MtlsConfig {
+  pub fn from_pem_buffers(certs: &[u8], key: &[u8]) -> Result<Self, String> {
+    let cert_chain = rustls_pemfile::certs(&mut std::io::Cursor::new(certs))
+      .map_err(|e| format!("Failed to read mTLS certs file: {e:?}"))?
+      .into_iter()
+      .map(rustls::Certificate)
+      .collect();
+
+    let keys = rustls_pemfile::read_all(&mut std::io::Cursor::new(key))
+      .map_err(|e| format!("Failed to read mTLS key file: {e:?}"))?
+      .into_iter()
+      .filter(|item| match item {
+        rustls_pemfile::Item::X509Certificate(_) => {
+          log::warn!("Found x509 certificate in mTLS key file. Ignoring.");
+          false
+        }
+        _ => true,
+      });
+
+    let mut key = None;
+    for item in keys {
+      use rustls_pemfile::Item;
+
+      match item {
+        Item::RSAKey(buf) | Item::PKCS8Key(buf) | Item::ECKey(buf) => {
+          key = Some(rustls::PrivateKey(buf))
+        }
+        Item::X509Certificate(_) => unreachable!("filtered above"),
+        _ => todo!("non-exhaustive match"),
```

Comment:
Should probably be a more useful end-user message?

Reply:
There's no other options, this is because the rustls_pemfile item enum is non-exhaustive. I'll change it to unreachable. Unfortunately since the items are `Item::Foo(Vec<u8>)` printing *what* was found would dump raw bytes to the console.